linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -3911,9 +3936,33 @@ private boolean advance() {
         }
         rights.add(right);
       }
-      cartesians = Linq4j.product(
-          ImmutableList.of(Linq4j.enumerator(lefts),
-              Linq4j.enumerator(rights)));
+
+      if (extraPredicate == null) {
+        cartesians = Linq4j.product(
+            ImmutableList.of(Linq4j.enumerator(lefts),
+                Linq4j.enumerator(rights)));
+      } else {
+        // we must verify the non equi-join predicate
+        final List<List<Object>> results = new ArrayList<>();
+        for (TSource currentLeft : lefts) {
+          for (TInner currentRight : rights) {
+            if (extraPredicate.apply(currentLeft, currentRight)) {
+              final List<Object> result = new ArrayList<>(2);
+              result.add(currentLeft);
+              result.add(currentRight);
+              results.add(result);
+            }
+          }
+        }
+        if (results.isEmpty()) {
+          if (done) {
+            return false;
+          }
+          return advance();
+        } else {
+          cartesians = Linq4j.enumerator(results);
```

Comment:
How is this different from a filter on top of `Linq4j.product`?

If it is not different, then please remove manual for-for.
If it is different please add a clarification comment.

Reply:
At the current state (only INNER join supported), we could have a filter on top of the product. However, we have other join types in the backlog that can / will be supported by EnumerableMergeJoin (SEMI-join, see [CALCITE-3833](https://issues.apache.org/jira/browse/CALCITE-3833); and ANTI-join, see  [CALCITE-3834](https://issues.apache.org/jira/browse/CALCITE-3834)). For those join types (which do not project the right-hand-side of the join), we cannot use the strategy of applying a filter on top of the product, so I am just preparing the ground here with an homogeneous strategy that can work on other join types.

Reply:
Implementation-wise it looks the same.
I see no reason why do you use `for-for-if` here.

Reply:
The next step will be implementing SEMI join type with Enumerable Merge Join. In order to do so, we would just a need a `break` inside the `if` (to avoid 'left' duplicates, which SEMI join must not have). So this algorithm works for INNER join and can be easily adapted for other types, like SEMI.

Reply:
Is it different from the existing `org.apache.calcite.linq4j.EnumerableDefaults#semiJoin` and `#antiJoin` methods?

Why do you reimplement the same thing?

Reply:
`org.apache.calcite.linq4j.EnumerableDefaults#semiJoin` and `#antiJoin` methods implement hashSemiJoin and hashAntiJoin, they are used by EnumerableHashJoin, the idea is to provide another alternative implementation of semiJoin/antiJoin via MergeJoin algorithm (as they exist for NestedLoopJoin, CorrelateJoin or CorrelateBatchJoin)

Reply:
Let me rephrase it differently:

1) You calculate `lefts` and `rights` using merge join algorithm on `equal` predicates
2) Then you have `lefts` and `rights`, and there's nothing from `merge` join left that you can do about it. The rest could be exactly the same as in `hash semi-join`.

`for-for-if` logic performs exactly the same actions as the existing `semiJoin` and `antoJoin` methods. Why duplicate code (bugs)?

Reply:
Ok, I think I understand your point. We should re-use existing code instead of the `for-for-if`. However, I don't think hashJoin code is the appropriate module here, since there is no need to create the hash structure to find matching keys (because we already have them with the merge join algorithm). Instead, we should use the `nestedLoopJoin(lefts, rights, extraPredicateWithNonEquiClauses)` which performs precisely this `for-for-if` process.

Reply:
> I don't think hashJoin code is the appropriate module here, since there is no need to create the hash structure to find matching keys

Well, my point is the code seems to duplicate the existing code, so it looks like a bug.
Of course, the only predicates left are `non-equality`, so there's nothing to hash. 

However, the key point here is to avoid code duplication.

Reply:
Refactored to re-use nested loop join